Fix bug in `visitDivExpr`, `visitModExpr` and `visitMulExpr`

arnab-polymage opened the pull request:
Whenever the result of a div, mod or mul affine expression is a constant expression, place the value in the constant index of the flattened expression instead of adding it as a local expression.

bondhugula commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -1178,9 +1178,9 @@ static AffineExpr getSemiAffineExprFromFlatForm(ArrayRef<int64_t> flatExprs,
-    if (!binaryExpr)
-      continue;
-
+    assert(binaryExpr &&
```
Instead of dyn_cast + assert, use `cast` and add the assertion message as a comment on top.

bondhugula replied:
Yes, that's exactly what I meant. The assertion message (string) can be added as a comment so developers know better the reason for using `cast`.

bondhugula reviewed (changes requested):
The changes aren't complete or sound.

CoTinker commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -1348,6 +1348,11 @@ LogicalResult SimpleAffineExprFlattener::visitModExpr(AffineBinaryOpExpr expr) {
     AffineExpr divisorExpr = getAffineExprFromFlatForm(rhs, numDims, numSymbols,
                                                        localExprs, context);
     AffineExpr modExpr = dividendExpr % divisorExpr;
+    if (auto constModExpr = dyn_cast<AffineConstantExpr>(modExpr)) {
```
What about `visitMulExpr`? Maybe it has same problem.

bondhugula commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -1177,10 +1177,9 @@ static AffineExpr getSemiAffineExprFromFlatForm(ArrayRef<int64_t> flatExprs,
-    if (!binaryExpr)
-      continue;
-
+    // Local expression cannot be a dimension, symbol or a constant -- it
```
A local expression ...

bondhugula reviewed (commented):
Missing test cases for the scenarios exercised. Changes are completely untested!

bondhugula commented:
> This is a stupid question. Will `semi affine expr` also have this constrain? If that's the case, the [fix](https://github.com/llvm/llvm-project/pull/145162) should be reverted.

Which constraint?

CoTinker commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -1295,7 +1294,23 @@ LogicalResult SimpleAffineExprFlattener::visitMulExpr(AffineBinaryOpExpr expr) {
     AffineExpr b = getAffineExprFromFlatForm(rhs, numDims, numSymbols,
                                              localExprs, context);
-    return addLocalVariableSemiAffine(mulLhs, rhs, a * b, lhs, lhs.size());
+    AffineExpr mulExpr = a * b;
```
The code in the three place is repeated，maybe a public function could be extracted.

lipracer commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -19,6 +19,7 @@
 #include "mlir/Support/TypeID.h"
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/Support/MathExtras.h"
+#include "llvm/Support/raw_ostream.h"
```
if not used, consider removing the include.

bondhugula commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -1295,7 +1315,8 @@ LogicalResult SimpleAffineExprFlattener::visitMulExpr(AffineBinaryOpExpr expr) {
                                              localExprs, context);
-    return addLocalVariableSemiAffine(mulLhs, rhs, a * b, lhs, lhs.size());
+    AffineExpr mulExpr = a * b;
+    return addExprToFlattenedList(mulLhs, rhs, lhs, mulExpr);
```
Combine into one line?

bondhugula commented on mlir/include/mlir/IR/AffineExprVisitor.h:
```diff
@@ -418,6 +418,14 @@ class SimpleAffineExprFlattener
 private:
+  /// Flatten binary expression `expr` and add it to `result`. If `expr` is a
+  /// dimension, symbol or constant, we add it to appropriate index in `result`.
+  /// Otherwise we add it in the local variable section. `lhs` and `rhs` are the
```
It's unclear here what operands of `expr` are! An AffineExpr doesn't have any operands. Rephrase.

bondhugula commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -1274,6 +1273,27 @@ SimpleAffineExprFlattener::SimpleAffineExprFlattener(unsigned numDims,
   operandExprStack.reserve(8);
 }
 
+LogicalResult SimpleAffineExprFlattener::addExprToFlattenedList(
```
I think this is completely misnamed. Is this specific to semi-affine exprs? In fact, addLocalIdSemiAffine in any of its overrides doesn't even use `lhs` and `rhs`. So all of these arguments aren't making sense to me.

```
LogicalResult SimpleAffineExprFlattener::addLocalIdSemiAffine(
    ArrayRef<int64_t> lhs, ArrayRef<int64_t> rhs, AffineExpr localExpr) {
  for (SmallVector<int64_t, 8> &subExpr : operandExprStack)
    subExpr.insert(subExpr.begin() + getLocalVarStartIndex() + numLocals, 0);
  localExprs.push_back(localExpr);
  ++numLocals;
  // lhs and rhs are not used here; an override of this method uses them.
  return success();
}
```

bondhugula replied:
Actually, I don't see the override using them either.

bondhugula commented on mlir/test/Dialect/Affine/simplify-structures.mlir:
```diff
@@ -608,3 +608,32 @@ func.func @semiaffine_simplification_floordiv_and_ceildiv_const(%arg0: tensor<?x
+// CHECK-DAG: #[[$MAP:.*]] = affine_map<()[s0] -> (13 mod s0)>
+// CHECK-DAG: #[[$MAP1:.*]] = affine_map<(d0) -> (d0 * 2)>
+// CHECK-LABEL: semiaffine_simplification_local_expr_folded_into_non_binary_expr
+func.func @semiaffine_simplification_local_expr_folded_into_non_binary_expr(%arg0: memref<?x?xf32>) -> (index, index) {
```
semi_affine

arnab-polymage committed:
Fix bug in `visitDivExpr`, `visitMulExpr`  and `visitModExpr`

Whenever the result of a div or mod affine expression is a
constant expression, place the value in the constant index of the
flattened expression instead of adding it as a local expression.

bondhugula commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -1274,6 +1273,27 @@ SimpleAffineExprFlattener::SimpleAffineExprFlattener(unsigned numDims,
+    result[getSymbolStartIndex() + symExpr.getPosition()] = 1;
+    return success();
+  }
+  return addLocalVariableSemiAffine(lhs, rhs, expr, result, result.size());
```
How do we know that `expr` is semi-affine at this stage? It can just be a purely affine binary expression. This method is confusing without any further comments. At this point, all you know is that `expr` is an affine binary expression and you might as well `cast` it to that and send it to make the signature of addLocal... less confusing.
